### PR TITLE
MESH-192 Bulk link api for domain<>asset link

### PR DESCRIPTION
**Pull request by PRATHAM2002-DS**
## Change description

> Created a custom API with minimum validations for bulk link domains usecase.

## Testcase Doc

[https://docs.google.com/spreadsheets/d/1I3JJ4gp_JUMcc4KXSX57RTZM-23ANu06SvDHQ-vDGpE/edit?gid=0#gid=0](url)

## Type of change
- [ ] Bug fix (fixes an issue)
- [ ] New feature (adds functionality)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [ ] Security impact of change has been considered
- [ ] Code follows company security practices and guidelines

### Code review 

- [ ] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [ ] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable

### CODE REVIEW

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java [hunk 1]
```diff
@@ -4628,4 +4679,14 @@ private void cacheDifferentialEntity(AtlasVertex ev, Set<String> existingValues)
...
+    private void cacheDifferentialMeshEntity(AtlasVertex ev, Set<String> existingValues) {
+        AtlasEntity diffEntity = new AtlasEntity(ev.getProperty(TYPE_NAME_PROPERTY_KEY, String.class));
+        diffEntity.setGuid(ev.getProperty(GUID_PROPERTY_KEY, String.class));
+        diffEntity.setAttribute(DOMAIN_GUIDS_ATTR, existingValues);
+        diffEntity.setUpdatedBy(ev.getProperty(MODIFIED_BY_KEY, String.class));
+        diffEntity.setUpdateTime(new Date(RequestContext.get().getRequestTime()));
+
+        RequestContext requestContext = RequestContext.get();
+        requestContext.cacheDifferentialEntity(diffEntity);
+    }
```
Comment (nikhilbonte21) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java:4703]: Please create a new helper method which will do the common part in this as well as previous mentod `cacheDifferentialEntity` 
New method will accept diffEntity & ev, set common attributes on diffEntity & add it into requestContext

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/AtlasEntityStoreV2.java [hunk 2]
```diff
@@ -2815,14 +2816,60 @@ public void unlinkBusinessPolicy(String policyGuid, Set<String> unlinkGuids) thr
...
+    public void linkMeshEntityToAsset(String meshEntityGuid, Set<String> linkGuids) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metric = RequestContext.get().startMetricRecord("linkProductToAsset.GraphTransaction");
+
+        try {
+            List<AtlasVertex> vertices = this.entityGraphMapper.linkMeshEntityToAsset(meshEntityGuid, linkGuids);
+            if (CollectionUtils.isEmpty(vertices)) {
+                return;
+            }
+
+            LOG.info("linkMeshEntityToAsset: entityGuid={}, linkGuids={}", meshEntityGuid, linkGuids);
```
Comment (nikhilbonte21) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/AtlasEntityStoreV2.java:2830]: Do not print whole linkGuids here, it will flood logs, just linkGuids.size() will be fine in case we want visibility

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java [hunk 3]
```diff
@@ -154,22 +154,22 @@
...
+    private static final Logger LOG = LoggerFactory.getLogger(EntityGraphMapper.class);
     private static final Logger PERF_LOG = AtlasPerfTracer.getPerfLogger("entityGraphMapper");
 
-    private static final String  SOFT_REF_FORMAT                   = "%s:%s";
-    private static final int     INDEXED_STR_SAFE_LEN              = AtlasConfiguration.GRAPHSTORE_INDEXED_STRING_SAFE_LENGTH.getInt();
-    private static final boolean WARN_ON_NO_RELATIONSHIP           = AtlasConfiguration.RELATIONSHIP_WARN_NO_RELATIONSHIPS.getBoolean();
-    private static final String  CUSTOM_ATTRIBUTE_KEY_SPECIAL_PREFIX = AtlasConfiguration.CUSTOM_ATTRIBUTE_KEY_SPECIAL_PREFIX.getString();
+    private static final String SOFT_REF_FORMAT = "%s:%s";
+    private static final int INDEXED_STR_SAFE_LEN = AtlasConfiguration.GRAPHSTORE_INDEXED_STRING_SAFE_LENGTH.getInt();
+    private static final boolean WARN_ON_NO_RELATIONSHIP = AtlasConfiguration.RELATIONSHIP_WARN_NO_RELATIONSHIPS.getBoolean();
```
Comment (nikhilbonte21) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java:162]: please revert these changes, not able to find the right code to review

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/MeshEntityNotifierImpl.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import org.apache.atlas.exception.AtlasBaseException;
+import org.apache.atlas.listener.EntityChangeListenerV2;
+import org.apache.atlas.model.instance.AtlasEntity;
+import org.apache.atlas.repository.graphdb.AtlasVertex;
+import org.apache.atlas.utils.AtlasPerfMetrics;
+import org.springframework.stereotype.Component;
+
+import javax.inject.Inject;
+import java.util.*;
+
```
Comment (nikhilbonte21) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/MeshEntityNotifierImpl.java:13]: can we reuse BusinessPolicyNotifierImpl?
I suggest renaming BusinessPolicyNotifierImpl to MinimalNotifierImpl

File: webapp/src/main/java/org/apache/atlas/web/rest/MeshEntityAssetLinkREST.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    /**
+     * Links a product to entities.
+     *
+     * @param domainGuid the ID of the product to be linked
+     * @param request    the request containing the GUIDs of the assets to link the product to
+     * @throws AtlasBaseException if there is an error during the linking process
+     */
+
+    @POST
+    @Path("/{domainId}/link-domain-to-asset")
```
Comment (nikhilbonte21) [inline webapp/src/main/java/org/apache/atlas/web/rest/MeshEntityAssetLinkREST.java:46]: Suggestion
link-domain-to-asset -> link-domain
unlink-domain-from-asset -> unlink-domain

File: webapp/src/main/java/org/apache/atlas/web/rest/MeshEntityAssetLinkREST.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    @Timed
+    public void linkDomainToAsset(@PathParam("domainId") final String domainGuid, final LinkMeshEntityRequest request) throws AtlasBaseException {
```
Comment (nikhilbonte21) [inline webapp/src/main/java/org/apache/atlas/web/rest/MeshEntityAssetLinkREST.java:48]: @PathParam("domainId") -> @PathParam("domainGuId")

Comment (nikhilbonte21) [inline webapp/src/main/java/org/apache/atlas/web/rest/MeshEntityAssetLinkREST.java:48]: linkDomainToAsset -> linkDomainToAssets

File: webapp/src/main/java/org/apache/atlas/web/rest/MeshEntityAssetLinkREST.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+            if (AtlasPerfTracer.isPerfTraceEnabled(PERF_LOG)) {
+                perf = AtlasPerfTracer.getPerfTracer(PERF_LOG, "MeshEntityAssetLinkREST.linkMeshEntityToAsset(" + domainGuid + ")");
+            }
+
+            // Link the domain to the specified entities
+            entitiesStore.linkMeshEntityToAsset(domainGuid, request.getLinkGuids());
+        } finally {
+            // Log performance metrics
+            AtlasPerfTracer.log(perf);
+            RequestContext.get().endMetricRecord(metric);
```
Comment (nikhilbonte21) [inline webapp/src/main/java/org/apache/atlas/web/rest/MeshEntityAssetLinkREST.java:74]: We have perf log added, remove metric log

File: webapp/src/main/java/org/apache/atlas/web/rest/MeshEntityAssetLinkREST.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+        AtlasPerfTracer perf = null;
+        try {
+            // Start performance tracing if enabled
+            if (AtlasPerfTracer.isPerfTraceEnabled(PERF_LOG)) {
+                perf = AtlasPerfTracer.getPerfTracer(PERF_LOG, "MeshEntityAssetLinkREST.unlinkDomainFromAsset(" + domainGuid + ")");
+            }
+
+            // Unlink the domain from the specified entities
+            entitiesStore.unlinkMeshEntityFromAsset(domainGuid, request.getUnlinkGuids());
```
Comment (nikhilbonte21) [inline webapp/src/main/java/org/apache/atlas/web/rest/MeshEntityAssetLinkREST.java:110]: unlinkMeshEntityFromAsset -> unlinkMeshEntity OR unlinkDomain for now

**Review by nikhilbonte21 — CHANGES_REQUESTED** [review]

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/MeshEntityNotifierImpl.java (hunk 4 above)
Comment (PRATHAM2002-DS) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/MeshEntityNotifierImpl.java:13]: Arpit had asked me to keep it seperate but even I think we can have one generic implementation. Will discuss with Arpit once again

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java [hunk 9]
```diff
@@ -4654,17 +4642,116 @@ public List<AtlasVertex> unlinkBusinessPolicy(String policyId, Set<String> unlin
...
+                }
+            }
+        }
+
+        return unlinkedVertices;
+    }
+
+    private void updateDomainAttribute(AtlasVertex vertex, Set<String> existingValues, String meshEntityId){
+        existingValues.forEach(existingValue -> vertex.removePropertyValue(DOMAIN_GUIDS_ATTR, existingValue));
+        vertex.setProperty(DOMAIN_GUIDS_ATTR, meshEntityId);
```
[the remark this answers is not in the document]
Comment (PRATHAM2002-DS) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java:4718]: Its added at the start no need to add it here

File: webapp/src/main/java/org/apache/atlas/web/rest/MeshEntityAssetLinkREST.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+        AtlasPerfTracer perf = null;
+        try {
+            // Start performance tracing if enabled
+            if (AtlasPerfTracer.isPerfTraceEnabled(PERF_LOG)) {
+                perf = AtlasPerfTracer.getPerfTracer(PERF_LOG, "MeshEntityAssetLinkREST.linkMeshEntityToAssets(" + domainGuid + ")");
+            }
+
+            // Link the domain to the specified entities
+            entitiesStore.linkMeshEntityToAssets(domainGuid, request.getAssetGuids());
```
[the remark this answers is not in the document]
Comment (PRATHAM2002-DS) [reply webapp/src/main/java/org/apache/atlas/web/rest/MeshEntityAssetLinkREST.java:68]: This API is triggered via Playbooks, so did not add this validation.

File: webapp/src/main/java/org/apache/atlas/web/rest/MeshEntityAssetLinkREST.java [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+        AtlasPerfTracer perf = null;
+        try {
+            // Start performance tracing if enabled
+            if (AtlasPerfTracer.isPerfTraceEnabled(PERF_LOG)) {
+                perf = AtlasPerfTracer.getPerfTracer(PERF_LOG, "MeshEntityAssetLinkREST.unlinkDomainFromAssets(" + domainGuid + ")");
+            }
+
+            // Unlink the domain from the specified entities
+            entitiesStore.unlinkMeshEntityFromAssets(domainGuid, request.getAssetGuids());
```
[the remark this answers is not in the document]
Comment (PRATHAM2002-DS) [reply webapp/src/main/java/org/apache/atlas/web/rest/MeshEntityAssetLinkREST.java:102]: This API is triggered via Playbooks, so did not add this validation.

**Review by nikhilbonte21 — APPROVED** [review]
